packages/portfolio-contract/src/portfolio.exo.ts

```diff
@@ -1,43 +1,192 @@
 
-type Accounts = {
-  Aave: never; // TODO
-  Compound: never; // TODO
-  USDN: OrchestrationAccount<{ chainId: 'noble-any' }>;
+const KeeperI = M.interface('keeper', {
+  init: M.call(M.string(), M.remotable('OrchestrationAccount')).returns(),
+  getAccount: M.call(M.string()).returns(M.remotable('OrchestrationAccount')),
+});
+
+const EvmTapI = M.interface('EvmTap', {
+  receiveUpcall: M.call(M.record()).returns(M.or(VowShape, M.undefined())),
+});
+
+const supportedDestinationChains: SupportedDestinationChains[] = [
+  'Avalanche',
+  'Base',
+  'Ethereum',
+];
+
+type EVMProtocolState = {
+  localAccount?: OrchestrationAccount<{ chainId: 'agoric-any' }>;
+  remoteAccountAddress?: string;
+  evmChain?: string;
+  isActive: boolean;
+};
+
+type CosmosProtocolState = {
+  icaAccount?: OrchestrationAccount<{ chainId: 'noble-any' }>;
+  isActive: boolean;
+};
+
+type PortfolioKitStateShape = {
+  protocolStates: MapStore<
+    YieldProtocol,
+    CosmosProtocolState | EVMProtocolState
+  >;
 };
 
 export const preparePortfolioKit = (zone: Zone) =>
   zone.exoClassKit(
     'Portfolio',
-    interfaceTODO,
-    () =>
-      ({
-        USDN: undefined,
-      }) as Partial<Accounts>,
     {
+      keeper: KeeperI,
+      invitationMakers: M.interface('invitationMakers', {
+        // supplyToAave: M.call(M.record()).returns(M.promise()),
+        // borrowFromAave: M.call(M.record()).returns(M.promise()),
+        // supplyToCompound: M.call(M.record()).returns(M.promise()),
+        // borrowFromCompound: M.call(M.record()).returns(M.promise()),
+        // withdraw: M.call(M.record()).returns(M.promise()),
+      }),
+      tap: EvmTapI,
+    },
+
+    (): PortfolioKitStateShape => {
+      return harden({
+        protocolStates: zone.mapStore('protocolStates'),
+      });
+    },
+    {
+      tap: {
+        receiveUpcall(event: VTransferIBCEvent) {
+          trace('receiveUpcall', event);
+
+          const tx: FungibleTokenPacketData = JSON.parse(
+            atob(event.packet.data),
+          );
+
+          trace('receiveUpcall packet data', tx);
+          const memo: AxelarGmpIncomingMemo = JSON.parse(tx.memo);
+
+          if (
+            supportedDestinationChains.includes(
+              memo.source_chain as SupportedDestinationChains,
+            )
+          ) {
+            const payloadBytes = decodeBase64(memo.payload);
+            const [{ isContractCallResult, data }] = decodeAbiParameters(
+              [
+                {
+                  type: 'tuple',
+                  components: [
+                    { name: 'isContractCallResult', type: 'bool' },
+                    {
+                      name: 'data',
+                      type: 'tuple[]',
+                      components: [
+                        { name: 'success', type: 'bool' },
+                        { name: 'result', type: 'bytes' },
+                      ],
+                    },
+                  ],
+                },
+              ],
```

Comment:
this whole data structure is a constant, yes? Did you consider hoisting it to a module level constant?